README.md
# Agrocartt
It is regarding supply chain management for the post harvest losses of crop products in India. We are creating a decentralized blockchain powered hyperledger project for solving this issue. We are open to suggestions and ideas that can help us so that we can grow as emerging tech company and we are creating a new digital era for farmers. This repository is under Paytm build for India Initiative. We want to create decentralized application for Farmers and customers. It is a Pan India Project. Developers who have experience in Blockchain,Solidity Programming and Mobile devlelopment are most welcome. Last but not least, those who have knowledge about agriculture and business are also inivited to join.

[Contributing Guidlines](CONTRIBUTING.md)

   
   <p align="center">
    <img src="assets/agro.png" alt="logo" width="100"/> <br>
    Logo By Anurag One of the contributor
    </p>

# For Contributing to Android Project

1. Clone the Repository
2. From android studio select Open Project -> (Path To) AgroCartt App
3. It will automatically compile and run.

[Android App Details and ChangeLogs](AgroCarttApp/CHANGELOG.md)

**Note(To the developers working on Android Project)**:- _It is requested that all the changes made should be logged in the [CHANGELOG.md](AgroCarttApp/CHANGELOG.md), this helps fellow developers to know the state of the project. If possible provide comments in the code especially when importing an external library or defining a method that the developer thinks is difficult to understand. 
Regards, **The AgroCartt Team**_

Current State of the app:-

 <p align="center">
    <img src="assets/sample.gif" alt="app"/> <br>
    </p>

* For initiating a bug report please read [Bug Report Template](https://github.com/PaytmBuildForIndia/Agrocartt/blob/master/.github/ISSUE_TEMPLATE/bug_report.md)
* For initiating a Feature Request Please read [Feature Request Template](https://github.com/PaytmBuildForIndia/Agrocartt/blob/master/.github/ISSUE_TEMPLATE/feature_request.md)


AgroCarttApp/CHANGELOG.md
# Android App CHANGELOGS

### LOG 1. Added Start Screen                                                                                          (06/07/18) (Contributing Member:- Sankalp Chauhan)
Notes:-
1. 
 * Min SDK Targeted:- 19 {SDK 19 == Android KITKAT}
 * Target SDK Version:- 26 (Android Oreo)
2. 
 * USING GLIDE LIBRARY TO DISPLAY GIF:-
  Some people also use picasso but, glide is the latest library and,
  The way of loading an image in Glide is way faster than Picasso.
  For more details visit:-
  [https://github.com/bumptech/glide](https://github.com/bumptech/glide)
3. 
 * MultiThread
         * Assigned a Handler for the communication between the UI and Background thread in android in start_activity.class this thread can be used for connecting to the server and fetching HTTP requests, and will have no impact on Main Thread of the Application. Instead of using a seperate class I have used Anonymous Inner class for implementing the runnable interface because in this activity, another implementation of runnable is not required For understanding why I have implemented this please refer this article by me:-
[Multithreading Core Java For Android Developers](https://medium.com/@sankalpchauhan.me/core-java-for-android-developers-multithreading-e7ec7d53924c) 
4. The GIF I used can be replaced with appropriate GIF (Further Details are mentioned in the comments)

      * BUG FIXED(6/07/2018):
      
       BUG
       Multiple MainActivities were launching when the user changes the device orientation
       during the GIF display.
       FIX
       Changed the configuration of .startScreen to ensure that it does not restart on orientation change
       I could have disabled screen orientation altogether but it would create problem in TABLETS
-------------------------------------------------------------------------------------------------------------------------------------------------------

### LOG 2. Added Language Selection Screen and Improved Look and Feel                                                       1:00 AM (08/07/18) (Contributing Member:- Sankalp Chauhan)
Change Description:- 

1. Major Change (Added Language Selection Screen)
Added a languge screen when the user clicks on the language the language will get selected and stored in an xml format to a file "lang_info" in shared_prefs folder this information will be recieved at runtime and will be used to render the language of the whole app. This also sets a language flag to true in lang_info.xml file which is fetched at runtime and accordingly user will not see the language selection screen everytime the app opens, rather it will be seen only one time after installation.

2. Minor Changes (Improved Look and Feel)
* Matched Colors with the initiative's default colors.
* Added a dialog box on exit (onBackButtonPressed()) so that the user do not accidently close the app
* Minor Bug Fixes and Ripple Effect dependency for pre-lollipop devices (API less than 21) 
(Further Details are mentioned in the code comments)

---------------------------------------------------------------------------------------------------------------------------------

### LOG 3. Added Sign In Screen (Using Firebase), Drawer For Profile Screen                                                   12:45 AM (09/07/18) (Contributing Member:- Sankalp Chauhan)
Change Description:-

1. **Added Sign In Screen** User can now login via Google Sign In, this was implemented using firebase [Firebase Authentication Guide](https://firebase.google.com/docs/auth/)
2. **Added a slide Drawer** Right now it contains placeholders, The placeholders will be replaced by user info.

* Known Issues:- (Please update the section after solving the Issue with explainaition on how issue is resolved)

        Known Issues:-
        Right Now Sign in is only possible through Google Sign In however Facebook sign In and Phone OTP buttons are present but they are non-functional.


----------------------------------------------------------------------------------------------------------------------------------


### LOG 4. Added No Internet Connection Dialog                                                   11:51 PM (10/07/18) (Contributing Member:- Nikhil Dhiman)
Change Description:-
_Same As Title_		

------------------------------------------------------------------------------------------------------------------------------------


### LOG 5. Added Methods For Fetching and Storing User Details from Firebase and Display it on Profile Screen                                          03:30 PM (11/07/18) (Contributing Member:- Sankalp Chauhan)
Change Description:-
The user's data is stored in SharedPreferences on Login Screen and then populated on next activity i.e. Main Activity.
Now when the user opens the app second time the user will not have to login again and the details will already be fetched from the shared prefrences.

-------------------------------------------------------------------------------------------------------------------------------------


* SORRY NOTE

		Sorry for inconvinience:-
		One of the user's pointed this out, the app was non functional from 18th July to 26 July'18 this was due to improper resolving of merge conflict by me (i.e. Sankalp Chauhan) which left unusual charachters at the end of each xml file however this has been resolved as of today i.e. 26 July '18. To make sure this does not happen in future I have downloaded the GUI application to automatically handle merge conflicts, earlier this was done manually.


------------------------------------------------------------------------------------------------------------------------------
### LOG 6. Added Progress Bar and UI improvements                                                                                                  09:15 PM (26/07/18) (Contributing Member:- Sankalp Chauhan)
Change Description:-
* Progress Bar added on login screen and added Phone OTP button.

AgroCarttApp/app/src/test/java/in/agrocartt/agrocartt/ExampleUnitTest.java
package in.agrocartt.agrocartt;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() throws Exception {
        assertEquals(4, 2 + 2);
    }
}

AgroCarttApp/app/src/main/java/in/agrocartt/agrocartt/startScreen.java
package in.agrocartt.agrocartt;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.view.WindowManager;
import android.widget.ImageView;
//Glide Library Import
import com.bumptech.glide.Glide;

/**
 * Created by AgroCartt Team on 6/7/18.
 * @AgroCartt
 */

public class startScreen extends AppCompatActivity {
    public static int TIME_OUT_ASSIGNED = 4000;     //Change the value of variable for varying the timeout
    private ImageView imageView;
    private Boolean check;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.start_screen);
        //Full Screen
        getWindow().setFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN,
                WindowManager.LayoutParams.FLAG_FULLSCREEN);
        //Setting ImageView
        imageView = findViewById(R.id.org_gif);

        // USING GLIDE LIBRARY TO DISPLAY GIF
        //some people also use picasso but, glide is the latest library and,
        //The way of loading an image in Glide is way faster than Picasso.
        //For more details visit:-
        // https://github.com/bumptech/glide

        Glide.with(this).load(R.drawable.agro).into(imageView);

        /**
         * @MultiThread (A short Note)
         * Assigning a Handler for the communication between the UI and Background thread in android.
         * this thread can be used for connecting to the server and fetching HTTP requests, and will have
         * no impact on Main Thread of Application. Regards Sankalp.
         *
         * Instead of using a seperate class I have used Anonymous Inner class for implementing the
         * runnable interface because in this activity, another implementation of runnable is not required
         * For understanding why I have implemented this please refer this article by me:-
         * @https://medium.com/@sankalpchauhan.me/core-java-for-android-developers-multithreading-e7ec7d53924c
         */
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                //Checks whether the user selected the language previously if not it will start LanguageSetting else MainActivity
                if(getLanguageSet() == false) {
                    Intent i = new Intent(startScreen.this, LanguageSetting.class);
                    startActivity(i);
                }
                else{
                    Intent i = new Intent(startScreen.this, LoginActivity.class);
                    startActivity(i);
                }
                finish();
            }
        }, TIME_OUT_ASSIGNED);

    }

    /**
     *
     * @returns the language that was selected by user previously from the SharedPrefrences
     */
    public boolean getLanguageSet(){
        SharedPreferences sharePref = getSharedPreferences("lang_info", Context.MODE_PRIVATE);
        check = sharePref.getBoolean("languageSet", false);
        return check;
    }
}


AgroCarttApp/app/src/main/java/in/agrocartt/agrocartt/LoginActivity.java
package in.agrocartt.agrocartt;



import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.util.LogPrinter;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.Toast;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.FirebaseException;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;
import com.google.firebase.auth.PhoneAuthCredential;
import com.google.firebase.auth.PhoneAuthProvider;

import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class LoginActivity extends AppCompatActivity {
    //a constant for detecting the login intent result
    private static final int RC_SIGN_IN = 234;
    //private ImageView user_profile;

    //for getting user's data from firebaseAuth
    String user_photoUrl = "";
    String user_displayName = "";
    String user_email = "";

    //Tag for the logs
    private static final String TAG = "Login_Activity";

    //creating a GoogleSignInClient object
    GoogleSignInClient mGoogleSignInClient;

    //And also a Firebase Auth object
    FirebaseAuth mAuth;

    ProgressBar progressBar;

    //For Verification Number Screen
    @BindView(R.id.phoneVerify)
    Button verifyBTN;



    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
       setContentView(R.layout.activity_login);
        ButterKnife.bind(this);

       //Network Connection is checked here:
        if(!isConnected(LoginActivity.this)) buildDialog(LoginActivity.this).show();


        //Intialized the FirebaseAuth object
        mAuth = FirebaseAuth.getInstance();

       // user_profile = findViewById(R.id.firebase_userPhoto); (DEPRECEATED)

        //Added a GoogleSignInOptions object
        //And we need to build it as below
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();

        //Then we will get the GoogleSignInClient object from GoogleSignIn class
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        //Now we will attach a click listener to the sign_in_button
        //and inside onClick() method we are calling the signIn() method that will open
        //google sign in intent
        findViewById(R.id.sign_in_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                signIn();
            }
        });
        progressBar = findViewById(R.id.login_progress);

    }

    @Override
    protected void onResume() {
        super.onResume();
        if(!isConnected(LoginActivity.this)) buildDialog(LoginActivity.this).show();
    }


//Getting user details from Firebase and storing it to shared prefrences
    //--------------------------------------------------------------------------------------------------------------------

    public void firebase_fetch_details(){
       FirebaseUser user = mAuth.getCurrentUser();
       if(user !=null){
           user_photoUrl = user.getPhotoUrl().toString();
           user_displayName = user.getDisplayName();
           user_email = user.getEmail();

           Log.i("Firebase details", "Details are " + "URL: " + user_photoUrl + "Name: " + user_displayName + "Email: "+ user_email);
       }
       dataToSharedPrefrences(user_photoUrl,user_displayName,user_email);
    }

    public void dataToSharedPrefrences(String fireurl, String firedisplayName, String fireemail){
        SharedPreferences sharePref = getSharedPreferences("user_info", Context.MODE_PRIVATE);
        SharedPreferences.Editor shareEdit = sharePref.edit();
        shareEdit.putString("firebase_userPhotourl", fireurl);
        shareEdit.putString("firebase_userDisplayName", firedisplayName);
        shareEdit.putString("firebase_userEmail", fireemail);
        shareEdit.commit();
        Log.i("FIREBASE DATA", "Data stored in Shared Prefrences");
    }
    //Getting user details from Firebase and storing it to shared prefrences over--------------------------------------------------------------------------------------------------------------------






    /**
     * Checks if the user is connected to the internet
     * @param context
     * @return
     */
    //CHECK NETWORK STATE----------------------------------------------------------------------------------------------------------------------
    public boolean isConnected(Context context) {

        ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo netinfo = cm.getActiveNetworkInfo();
        Log.i("NetworkState", "isConnected checking network state.......");
        if (netinfo != null && netinfo.isConnectedOrConnecting()) {
            android.net.NetworkInfo wifi = cm.getNetworkInfo(ConnectivityManager.TYPE_WIFI);
            android.net.NetworkInfo mobile = cm.getNetworkInfo(ConnectivityManager.TYPE_MOBILE);

            if((mobile != null && mobile.isConnectedOrConnecting()) || (wifi != null && wifi.isConnectedOrConnecting())) return true;
        else return false;
        } else
        return false;
    }

    public AlertDialog.Builder buildDialog(Context c) {
        Log.i("NetworkState", "Network Failed......... Building Dialog");
        AlertDialog.Builder builder = new AlertDialog.Builder(c);
        builder.setTitle("No Internet Connection");
        builder.setMessage("You need to have Mobile Data or WIFI to access AgroCartt. Press Enable to access network settings");
        builder.setIcon(android.R.drawable.ic_dialog_alert);

        builder.setPositiveButton("Enable", new DialogInterface.OnClickListener() {

            @Override
            public void onClick(DialogInterface dialog, int which) {
                Intent in = new Intent(android.provider.Settings.ACTION_DATA_ROAMING_SETTINGS);
                startActivity(in);
               //finish();
            }
        });

        return builder;
    }
    //CHECK NETWORK STATE METHODS OVER----------------------------------------------------------------------------------------------------------------------





    @Override
    protected void onStart() {
        super.onStart();

        //if the user is already signed in
        //we will close this activity
        //and take the user to profile activity
        if (mAuth.getCurrentUser() != null) {
            finish();
            startActivity(new Intent(this, MainActivity.class));
        }
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        //if the requestCode is the Google Sign In code that we defined at starting
        if (requestCode == RC_SIGN_IN) {

            //Getting the GoogleSignIn Task
            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            try {
                //Google Sign In was successful, authenticate with Firebase
                GoogleSignInAccount account = task.getResult(ApiException.class);

                //authenticating with firebase
                firebaseAuthWithGoogle(account);
            } catch (ApiException e) {
                Log.v("SHA1 Failiure", e.getMessage());
                Toast.makeText(LoginActivity.this, e.getMessage(), Toast.LENGTH_SHORT).show();
                Toast.makeText(LoginActivity.this, "API Call FAIL", Toast.LENGTH_SHORT).show();
            }
        }
    }





    //firebaseAuthentication start----------------------------------------------------------------------------------------------------------------------

    private void firebaseAuthWithGoogle(GoogleSignInAccount acct) {

        progressBar.setVisibility(View.VISIBLE);
        Log.d(TAG, "firebaseAuthWithGoogle:" + acct.getId());

        //getting the auth credential
        AuthCredential credential = GoogleAuthProvider.getCredential(acct.getIdToken(), null);

        //Now using firebase we are signing in the user here
        mAuth.signInWithCredential(credential)
                .addOnCompleteListener(this, new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        progressBar.setVisibility(View.GONE);
                        if (task.isSuccessful()) {
                            Log.d(TAG, "signInWithCredential:success");
                            firebase_fetch_details();
                            Intent i = new Intent(LoginActivity.this, MainActivity.class);
                            startActivity(i);
                            Log.i("FIREBASE", "Sign In Success");
                            LoginActivity.this.finish();
                            Toast.makeText(LoginActivity.this, user_displayName + " Signed In", Toast.LENGTH_SHORT).show();
                        } else {
                            // If sign in fails, a message is displayed to the user.
                            Log.w(TAG, "signInWithCredential:failure", task.getException());
                            Toast.makeText(LoginActivity.this, "Authentication failed.",
                                    Toast.LENGTH_SHORT).show();

                        }

                    }
                });
    }


    //this method is called on click
    private void signIn() {
        //getting the google signin intent
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();

        //starting the activity for result
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    //firebaseAuthentication start methods over ----------------------------------------------------------------------------------------------------------------------






    //A simple back dialog as is used in MainActivity
    @Override
    public void onBackPressed() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage("This app requires sign in to work, Are you sure you want to exit?")
                .setCancelable(false)
                .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        LoginActivity.this.finish();
                        finish();
                    }
                })
                .setNegativeButton("No", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                    }
                });
        AlertDialog alert = builder.create();
        alert.show();
    }

    @OnClick(R.id.phoneVerify)
    void toPhoneVerifyScreen(View view) {
       Intent i = new Intent(this, PhoneVerifyActivity.class);
       startActivity(i);


    }





}













AgroCarttApp/app/src/main/java/in/agrocartt/agrocartt/LanguageSetting.java
package in.agrocartt.agrocartt;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;

public class LanguageSetting extends AppCompatActivity implements View.OnClickListener {
    private TextView lang1, lang2, lang3;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_language_setting);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Snackbar.make(view, "More Languages Coming Soon...", Snackbar.LENGTH_LONG)
                        .setAction("Action", null).show();
            }
        });
        lang1 = findViewById(R.id.eng);
        lang2 = findViewById(R.id.punjabi);
        lang3 = findViewById(R.id.hindi);
        lang1.setOnClickListener(this);
        lang2.setOnClickListener(this);
        lang3.setOnClickListener(this);

    }

    //Implementation of onClickListner
    @Override
    public void onClick(View view) {
        switch (view.getId()){
            case R.id.eng:{
                //TODO Logic here
                isLanguageSet(true,"english");
                moveToNextActivity();
                finish();
            }
            break;
            case R.id.punjabi:{
                //TODO Logic here
                Log.i("PUNJABI", "PUNJABI SELECTED");
                moveToNextActivity();
                isLanguageSet(true,"punjabi");
                finish();
            }
            break;
            case R.id.hindi:{
                //TODO Logic here
                Log.i("HINDI", "HINDI SELECTED");
                isLanguageSet(true, "hindi");
                moveToNextActivity();
                finish();
            }
            break;
            default:{
                Toast.makeText(LanguageSetting.this, "Please select a language to continue", Toast.LENGTH_LONG).show();
            }
        }

    }

    /**
     * Stores the state of the language setting in SharedPrefrences with the filename "lang_info.xml"
     * @param bool for storing the state of setting with key language set and value true or false
     * @param language for storing the language selected with key selectedLanguage and value = hindi, english,punjabi
     */
    public void isLanguageSet(Boolean bool, String language){
        SharedPreferences sharePref = getSharedPreferences("lang_info", Context.MODE_PRIVATE);
        SharedPreferences.Editor shareEdit = sharePref.edit();
        shareEdit.putBoolean("languageSet", bool);
        shareEdit.putString("selectedLanguage", language);
        shareEdit.commit();
    }


    // Sets an intent to move to next Activity
    public void moveToNextActivity(){
        Intent i = new Intent(LanguageSetting.this, LoginActivity.class);
        startActivity(i);
    }



}


AgroCarttApp/app/src/main/java/in/agrocartt/agrocartt/MainActivity.java
package in.agrocartt.agrocartt;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Bitmap;
import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * A small note:-
 *Min SDK Targeted:- 19 {SDK 19 == Android KITKAT}
 *Target SDK Version:- 26 (Android Oreo)
 *
 * Created by AgroCartt Team on 6/7/18.
 */


public class MainActivity extends AppCompatActivity {

    @BindView(R.id.drawer_layout) DrawerLayout mDrawerLayout;
    ImageView profileImage;
    TextView textName, textEmail;
    FirebaseAuth mAuth;

    //Getting Firebase data from shared Prefrences:-
    String firebase_UserPhotourl;
    String firebase_userDisplayName;
    String firebase_userEmail;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        mAuth = FirebaseAuth.getInstance();

        //-------------------------NAVIGATION HEADER---------------------------------------
        //Getting the views of navigaion header explicitly as mentioned in documentation
        View header = ((NavigationView)findViewById(R.id.nav_view)).getHeaderView(0);
        profileImage = header.findViewById(R.id.firebase_userPhoto);
        textName = header.findViewById(R.id.firebase_userName);
        textEmail = header.findViewById(R.id.firebase_userEmail);
        //FETCHING USER DATA FROM SHARED PREFRENCES (I guess there is no need to check whether the user is present or not because that check has been performed on previous page)

        getFirebaseUserFromSharedPrefrences();
        if(firebase_userDisplayName.length()>0 && firebase_userEmail.length()>0)
        {
            if (getFirebaseUserFromSharedPrefrences() == true) {
                //Setting Data of User:-
                textName.setText(firebase_userDisplayName);
                textEmail.setText(firebase_userEmail);
                //check condition before url is placed

                    Glide.with(this).load(firebase_UserPhotourl).into(profileImage);

            }
        }
        //----------------------------HEADER SECTION END-----------------------------------




        //----------------------------DRAWER SECTION---------------------------------------

        //Sets toolbar burger for Drawer

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionbar = getSupportActionBar();
        actionbar.setDisplayHomeAsUpEnabled(true);
        actionbar.setHomeAsUpIndicator(R.drawable.ic_menu_black_24dp);

        //mDrawerLayout = findViewById(R.id.drawer_layout);

        NavigationView navigationView = findViewById(R.id.nav_view);
        navigationView.setNavigationItemSelectedListener(
                new NavigationView.OnNavigationItemSelectedListener() {
                    @Override
                    public boolean onNavigationItemSelected(MenuItem menuItem) {
                        // set item as selected to persist highlight
                        menuItem.setChecked(true);
                        // close drawer when item is tapped
                        mDrawerLayout.closeDrawers();

                        // Add code here to update the UI based on the item selected
                        switch(menuItem.getItemId()){
                            case R.id.signOut:
                                Toast.makeText(MainActivity.this, "Thankyou" + textName, Toast.LENGTH_SHORT).show();
                                signOut();
                                return true;
                        }
                        // For example, swap UI fragments here

                        return true;
                    }
                });

        //----------------------------DRAWER SECTION END---------------------------------------



    }


    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                mDrawerLayout.openDrawer(GravityCompat.START);
                return true;
                //Don't add drawer items here it's in the drawer section.
        }
        return super.onOptionsItemSelected(item);
    }


    /**
     * A simple dialog is shown on back button pressed so that the user does not accidently
     * exits the app
     */
    @Override
    public void onBackPressed() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage("Are you sure you want to exit?")
                .setCancelable(false)
                .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        finish();
                    }
                })
                .setNegativeButton("No", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        dialog.cancel();
                    }
                });
        AlertDialog alert = builder.create();
        alert.show();
    }

    @Override
    protected void onStart() {
        super.onStart();

        //if the user is not logged in
        //opening the login activity
        if (mAuth.getCurrentUser() == null) {
            finish();
            startActivity(new Intent(this, LoginActivity.class));
        }
    }


    //Gets the user data from the shared prefrences this will be used to populate the drawer
    public boolean getFirebaseUserFromSharedPrefrences(){
        SharedPreferences sharePref = getSharedPreferences("user_info", Context.MODE_PRIVATE);
        firebase_UserPhotourl = sharePref.getString("firebase_userPhotourl","Check Internet Connection");
        firebase_userDisplayName = sharePref.getString("firebase_userDisplayName","Check Internet Connection");
        firebase_userEmail = sharePref.getString("firebase_userEmail","Check Internet Connection");

        Log.i("FIREBASEfetch", "Data is"+ firebase_UserPhotourl + firebase_userDisplayName + firebase_userEmail );

        return true;
    }


    //A function for sign out from firebase
    public void signOut(){
        try {
            FirebaseAuth.getInstance().signOut();
            Log.i("Sign Out: ", "SUCCESS");
            Intent i = new Intent(this, LoginActivity.class);
            startActivity(i);
        } catch (Exception e) {
            //Toast.makeText(this, ""+e.getMessage(), Toast.LENGTH_LONG).show();
            e.printStackTrace();
        }

    }


}

